python/ray/serve/tests/unit/test_deployment_state.py

```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
+        )
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.STARTING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.UPDATING
+
+        for replica in deployment_state._replicas.get():
+            replica._actor.set_ready()
+
+        deployment_state.update(target_capacity=50)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        deployment_state.update(target_capacity=50)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_target_capacity_100_no_effect(self, mock_deployment_state):
+        """
+        Deploy with no target_capacity set, then set to 100. Should take no effect.
+
+        Then go back to no target_capacity, should still have no effect.
+        """
+        deployment_state, timer, cluster_node_info_cache = mock_deployment_state
+
+        b_info_1, b_version_1 = deployment_info(num_replicas=2)
+        updating = deployment_state.deploy(b_info_1)
+        assert updating
+
+        # Initially deploy with no target_capacity set.
+        deployment_state_update_result = deployment_state.update()
+        deployment_state._deployment_scheduler.schedule(
+            {deployment_state._id: deployment_state_update_result.upscale}, {}
+        )
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.STARTING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.UPDATING
+
+        for replica in deployment_state._replicas.get():
+            replica._actor.set_ready()
+
+        deployment_state.update()
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        deployment_state.update()
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Now update target_capacity to 100, should have no effect.
+        deployment_state.update(target_capacity=100)
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Now update target_capacity back to None, should have no effect.
+        deployment_state.update(target_capacity=None)
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_target_capacity_0_1_replica(self, mock_deployment_state):
+        """
+        Deploy with target_capacity set to 0. Should have a single replica.
```

Comment:
If `target_capacity` is 0, I'd expect there to be 0 replicas because we'd be scaling `num_replicas` and the `autoscaling_bounds` down to 0. Is there a reason 0 is treated as a special case?

Reply:
This is to ensure our invariant that all requests can always succeed. I've fixed it for scale-to-zero, but for deployments with a fixed num_replicas, they should still always have at least one replica.

Reply:
Otherwise `target_capacity=0` is equivalent to just deleting all applications

Reply:
> Otherwise `target_capacity=0` is equivalent to just deleting all applications

I think that make sense. `target_capacity=0` is equivalent to scaling the target version to 0. Imo, the invariant should be that `target_capacity>0` means at least 1 replica for each (non-scale-to-zero) deployment should be running. And this implementation already supports that because it rounds up when doing `num_replicas` calculations.

Reply:
There's no point to ever using `target_capacity=0` then, because you might as well just delete everything (if scaling down) or make your first step `>0` (if scaling up).

Also, if this cannot be used to express the minimal scale of the applications, there's no other semantically clear way to do it. Do you pick `10`? Then if you have >10 replicas, it doesn't work. Same for `1` (`>100` replicas) or `0.1` (`>1000` replicas).

Makes things messy.

Reply:
> Also, if this cannot be used to express the minimal scale of the applications, there's no other semantically clear way to do it.

That's a good point I hadn't considered. I'm okay with keeping the invariant as-is: there's always at least 1 replica for every deployment.